In [1]:
import IPython
import nltk
from nltk.corpus import wordnet as wn
import pandas as pd
from tqdm import tqdm_notebook as tqdm

Выбираю отправной синсет

In [2]:
target = 'travel.v.01'
base = wn.synset(target)
syns = base.hyponyms()
syn_names = [s.name() for s in syns]
count = len(syn_names)
print(syn_names)

['accompany.v.02', 'advance.v.01', 'angle.v.01', 'ascend.v.01', 'automobile.v.01', 'back.v.02', 'bang.v.04', 'beetle.v.02', 'betake_oneself.v.01', 'billow.v.02', 'bounce.v.03', 'breeze.v.02', 'caravan.v.01', 'career.v.01', 'carry.v.36', 'circle.v.01', 'circle.v.02', 'circuit.v.01', 'circulate.v.07', 'come.v.01', 'come.v.11', 'crawl.v.01', 'cruise.v.02', 'derail.v.02', 'descend.v.01', 'do.v.13', 'drag.v.04', 'draw.v.12', 'drive.v.02', 'drive.v.14', 'ease.v.01', 'fall.v.01', 'fall.v.15', 'ferry.v.03', 'float.v.01', 'float.v.02', 'float.v.05', 'flock.v.01', 'fly.v.01', 'fly.v.06', 'follow.v.01', 'follow.v.04', 'forge.v.05', 'get_around.v.04', 'ghost.v.01', 'glide.v.01', 'go_around.v.02', 'hiss.v.02', 'hurtle.v.01', 'island_hop.v.01', 'lance.v.01', 'lurch.v.03', 'outflank.v.01', 'pace.v.02', 'pan.v.01', 'pass.v.01', 'pass_over.v.04', 'play.v.09', 'plow.v.03', 'prance.v.02', 'precede.v.04', 'precess.v.01', 'proceed.v.02', 'propagate.v.02', 'pursue.v.02', 'push.v.09', 'raft.v.02', 'repair.v.

In [3]:
target in syn_names

False

In [4]:
count

123

## Матрица

Создаю пустую матрицу

In [5]:
matrix = pd.DataFrame([[0]*count]*count, index=syn_names, columns=syn_names)

In [6]:
matrix.head()

accompany.v.02  advance.v.01  angle.v.01  ascend.v.01  \
accompany.v.02                0             0           0            0   
advance.v.01                  0             0           0            0   
angle.v.01                    0             0           0            0   
ascend.v.01                   0             0           0            0   
automobile.v.01               0             0           0            0   

                 automobile.v.01  back.v.02  bang.v.04  beetle.v.02  \
accompany.v.02                 0          0          0            0   
advance.v.01                   0          0          0            0   
angle.v.01                     0          0          0            0   
ascend.v.01                    0          0          0            0   
automobile.v.01                0          0          0            0   

                 betake_oneself.v.01  billow.v.02  ...  weave.v.04  wend.v.01  \
accompany.v.02                     0            0  ...           0          0   
advance.v.01                       0            0  ...           0          0   
angle.v.01                         0            0  ...           0          0   
ascend.v.01                        0            0  ...           0          0   
automobile.v.01                    0            0  ...           0          0   

                 wheel.v.03  whine.v.01  whish.v.02  whisk.v.02  whistle.v.02  \
accompany.v.02            0           0           0           0             0   
advance.v.01              0           0           0           0             0   
angle.v.01                0           0           0           0             0   
ascend.v.01               0           0           0           0             0   
automobile.v.01           0           0           0           0             0   

                 withdraw.v.01  zigzag.v.01  zoom.v.02  
accompany.v.02               0            0          0  
advance.v.01                 0            0          0  
angle.v.01                   0            0          0  
ascend.v.01                  0            0          0  
automobile.v.01              0            0          0  

[5 rows x 123 columns]

Складываю кол-во совпадающих лемм во всех языках

In [7]:
def count_lemmas(s1, s2):
    n = 0
    for lang in wn.langs():

        l1 = s1.lemmas(lang)
        l2 = s2.lemmas(lang)

        n +=  len(set(l1) & set(l2))
    return n

In [8]:
for s1 in tqdm(syns):
    for s2 in syns:
        
        if s1 == s2:
            continue
        
        n = count_lemmas(s1, s2)
        
        matrix[s1.name()][s2.name()] = n
        matrix[s2.name()][s1.name()] = n

In [9]:
matrix.head(10)

accompany.v.02  advance.v.01  angle.v.01  ascend.v.01  \
accompany.v.02                    0             0           0            0   
advance.v.01                      0             0           0            1   
angle.v.01                        0             0           0            0   
ascend.v.01                       0             1           0            0   
automobile.v.01                   0             0           0            0   
back.v.02                         0             0           0            0   
bang.v.04                         0             0           0            0   
beetle.v.02                       0             0           0            0   
betake_oneself.v.01               0             0           0            0   
billow.v.02                       0             0           0            0   

                     automobile.v.01  back.v.02  bang.v.04  beetle.v.02  \
accompany.v.02                     0          0          0            0   
advance.v.01                       0          0          0            0   
angle.v.01                         0          0          0            0   
ascend.v.01                        0          0          0            0   
automobile.v.01                    0          0          0            0   
back.v.02                          0          0          0            0   
bang.v.04                          0          0          0            0   
beetle.v.02                        0          0          0            0   
betake_oneself.v.01                0          0          0            0   
billow.v.02                        0          0          0            0   

                     betake_oneself.v.01  billow.v.02  ...  weave.v.04  \
accompany.v.02                         0            0  ...           0   
advance.v.01                           0            0  ...           0   
angle.v.01                             0            0  ...           0   
ascend.v.01                            0            0  ...           0   
automobile.v.01                        0            0  ...           0   
back.v.02                              0            0  ...           0   
bang.v.04                              0            0  ...           0   
beetle.v.02                            0            0  ...           0   
betake_oneself.v.01                    0            0  ...           0   
billow.v.02                            0            0  ...           0   

                     wend.v.01  wheel.v.03  whine.v.01  whish.v.02  \
accompany.v.02               0           0           0           0   
advance.v.01                 0           1           0           0   
angle.v.01                   0           0           0           0   
ascend.v.01                  0           0           0           0   
automobile.v.01              0           0           0           0   
back.v.02                    0           0           0           0   
bang.v.04                    0           0           0           0   
beetle.v.02                  0           0           0           0   
betake_oneself.v.01          2           0           0           0   
billow.v.02                  0           0           0           0   

                     whisk.v.02  whistle.v.02  withdraw.v.01  zigzag.v.01  \
accompany.v.02                0             0              0            0   
advance.v.01                  0             0              0            0   
angle.v.01                    0             0              0            0   
ascend.v.01                   0             0              0            0   
automobile.v.01               0             0              0            0   
back.v.02                     0             0             10            0   
bang.v.04                     0             0              0            0   
beetle.v.02                   0             0              0            0   
betake_oneself.v.01           0             0            

In [158]:
#matrix.to_csv('matrix.csv')

Так можно быстро посмотреть на связи любого синсета

In [12]:
def look(s_name):
    return matrix[matrix[s_name] != 0][s_name]

In [46]:
look('accompany.v.02')

follow.v.01     12
follow.v.04      3
precede.v.04     2
proceed.v.02     1
pursue.v.02      3
Name: accompany.v.02, dtype: int64

In [47]:
look('drive.v.02')

ascend.v.01      2
do.v.13          1
drive.v.14      10
precede.v.04     5
pursue.v.02      1
raft.v.02        1
return.v.01      1
ride.v.01        3
ride.v.04        2
ride.v.10        9
roll.v.12        2
taxi.v.01        1
travel.v.05      1
travel.v.06      1
wend.v.01        1
wheel.v.03       4
Name: drive.v.02, dtype: int64

## Анализ

In [15]:
import networkx as nx
import matplotlib.pyplot as plt

In [38]:
G = nx.from_pandas_adjacency(matrix)

In [39]:
#nx.write_gexf(G, 'nx_graph.gexf')

In [40]:
print(nx.info(G))
print('Number of connected components:', nx.number_connected_components(G))
print('Density', nx.density(G))
print('Assortativity coefficient:', nx.degree_pearson_correlation_coefficient(G))
print('Av. clustering coefficient:', nx.average_clustering(G))
print('Transitivity:', nx.transitivity(G))

Name: 
Type: Graph
Number of nodes: 123
Number of edges: 633
Average degree:  10.2927
Number of connected components: 26
Density 0.08436625349860057
Assortativity coefficient: 0.07961415763622556
Av. clustering coefficient: 0.43135884355603743
Transitivity: 0.4759366671892146


### еще статистика

In [213]:
IPython.display.HTML(filename='./path/report.html')

![](degree/degree-distribution.png)

#### weighted

![](./w_degree/w-degree-distribution.png)

### визуализация

Мне не очень понравилось, что получалось при Girvan-Newman кластеризации:

<img src="newman.png" width="600">

In [210]:
IPython.display.HTML(filename='./newman/report.html')

пэтому я остновилась на Modularity class (цвет) и Betweeness Centrality (размер). **Вот итоговая визуализация:**

![](betwcentr_size_modularity_color.png)

*(Я не очень поняла, почему greedy_modularity в networkx дает какой-то сильно другой результат, поэтому дополнительно привожу k_clique и asyn_fluidc)*

In [43]:
from networkx import community

In [99]:
coms = community.k_clique_communities(G, 15)
for com in coms:
    if len(com) > 1:
        print(', '.join(com), end='\n\n')

In [89]:
largest = max(nx.connected_component_subgraphs(G), key=len)
coms2 = community.asyn_fluidc(largest, 11)

for com in coms2:
    if len(com) > 1:
        print(', '.join(com), end='\n\n')

ferry.v.03, pass.v.01, pass_over.v.04, transfer.v.06, sift.v.01

repair.v.03, wheel.v.03, pan.v.01, ride.v.10, turn.v.06, swing.v.03, weave.v.04, roll.v.12, stray.v.02

fly.v.01, ascend.v.01, swim.v.01, ride.v.01, float.v.02, swim.v.05, float.v.05, travel_by.v.01, float.v.01

rush.v.01, hurtle.v.01, lance.v.01, career.v.01, scramble.v.01, speed.v.04, travel_rapidly.v.01, whisk.v.02, forge.v.05

round.v.01, circle.v.02, circuit.v.01, travel.v.06, outflank.v.01, tram.v.01, circle.v.01

follow.v.04, pursue.v.02, crawl.v.01, proceed.v.02, follow.v.01, accompany.v.02, trail.v.03, drag.v.04

go_around.v.02, get_around.v.04, fly.v.06, circulate.v.07, propagate.v.02, cruise.v.02

walk.v.10, walk.v.01, pace.v.02, betake_oneself.v.01, ride.v.04, wend.v.01, step.v.02, retrograde.v.02, retreat.v.02, travel_purposefully.v.01, withdraw.v.01, shuttle.v.01, travel.v.05, step.v.01, travel.v.04, back.v.02, step.v.06

do.v.13, run.v.34, advance.v.01, descend.v.01, come.v.01, fall.v.01, carry.v.36, fall.v

### Центральность

Топ по degree:

In [23]:
deg = nx.degree_centrality(G)
for node in sorted(deg, key=deg.get, reverse=True)[:12]:
    print(node)

roll.v.12
travel_by.v.01
advance.v.01
travel.v.05
do.v.13
ride.v.10
travel.v.06
come.v.01
proceed.v.02
ride.v.01
rush.v.01
walk.v.01


<img src="top_degree.png" width="600">

Топ по betweenness centrality:

In [25]:
b = nx.betweenness_centrality(G)
for node in sorted(b, key=b.get, reverse=True)[:11]:
    print(node)

rush.v.01
roll.v.12
advance.v.01
proceed.v.02
pass.v.01
travel.v.05
wheel.v.03
travel_by.v.01
weave.v.04
travel.v.06
career.v.01


<img src="top_betw.png" width="600">

Топ по eigencentrality:

In [27]:
e = nx.eigenvector_centrality(G)
for node in sorted(b, key=b.get, reverse=True)[:12]:
    print(node)

rush.v.01
roll.v.12
advance.v.01
proceed.v.02
pass.v.01
travel.v.05
wheel.v.03
travel_by.v.01
weave.v.04
travel.v.06
career.v.01
ride.v.10


<img src="eigen_80.png" width="600">

Топ по closeness centrality:

In [31]:
c = nx.closeness_centrality(G)
for node in sorted(c, key=b.get, reverse=True)[:14]:
    print(node)

rush.v.01
roll.v.12
advance.v.01
proceed.v.02
pass.v.01
travel.v.05
wheel.v.03
travel_by.v.01
weave.v.04
travel.v.06
career.v.01
ride.v.10
draw.v.12
do.v.13


<img src="clos_55.png" width="600">

## Выводы

26 компонент, из которых одна состоит из 2 элементов (*zoom.v.02* и *whish.v.02*) и 24 одинарных:

In [221]:
print([i for i in pmatrix if sum(pmatrix[i]) == 0])

['angle.v.01', 'automobile.v.01', 'bang.v.04', 'beetle.v.02', 'billow.v.02', 'bounce.v.03', 'caravan.v.01', 'derail.v.02', 'ease.v.01', 'flock.v.01', 'ghost.v.01', 'island_hop.v.01', 'lurch.v.03', 'plow.v.03', 'prance.v.02', 'precess.v.01', 'seek.v.04', 'slice_into.v.01', 'snowshoe.v.01', 'steamer.v.01', 'swap.v.02', 'swash.v.01', 'trundle.v.01', 'whine.v.01']


*roll.v.12* всегда высоко топе по всем приведенным метрикам центральности. Я, честно говоря, не совсем понимаю почему, но предположу, что причина в том, что это значение хуже всех лексикализовано, поэтому у него очень много вариантов выражения лексемами с другими основными значениями.

In [122]:
wn.synset('roll.v.12').definition()

'move about aimlessly or without any destination, often in search of food or employment'

In [123]:
look('roll.v.12')

advance.v.01            1
carry.v.36              1
circle.v.02             2
circuit.v.01            1
circulate.v.07          2
come.v.01               1
come.v.11               1
descend.v.01            1
do.v.13                 3
draw.v.12               3
drive.v.02              2
fall.v.01               1
fall.v.15               2
float.v.01             24
float.v.02              4
fly.v.01                2
get_around.v.04         2
glide.v.01              2
go_around.v.02          6
outflank.v.01           2
pan.v.01                1
proceed.v.02            1
repair.v.03             1
ride.v.01               1
ride.v.10               1
rise.v.01               1
run.v.11                2
run.v.34                1
rush.v.01               1
step.v.02               2
stray.v.02             23
swim.v.01               2
swing.v.03              1
transfer.v.06           2
travel.v.04             8
travel.v.05             5
travel.v.06             2
travel_rapidly.v.01     1
turn.v.06   

Еще "важными" оказываются:
*rush.v.01, advance.v.01, proceed.v.02, pass.v.01* и *travel.v.05*. Если посмотреть на их дефиниции, можно сказать, что это самые "широкие" значения (наименее специфичные):

In [121]:
for s in 'rush.v.01, advance.v.01, proceed.v.02, pass.v.01, travel.v.05'.split(', '):
    d = wn.synset(s).definition()
    print(f'{s}:\t{d}')

rush.v.01:	move fast
advance.v.01:	move forward, also in the metaphorical sense
proceed.v.02:	move ahead; travel onward in time or space
pass.v.01:	go across or through
travel.v.05:	undergo transportation as in a vehicle


При разделении на сообщества, можно выделить следующие относительно устойчивые кластеры:
* манера передвижения (roll, wheel, weave, zigzag)

* передвижение по окружности (go around, circulate, circle)

* передвижение с кем-то / следование (follow, trail, accompany)

* движение вперед (proceed, advance)

* движение назад (back, withdraw, retreat)

* быстрое движение (speed, rush, travel_rapidly, career, whisk, hiss, lance)

* инструмент передвижения (step, ride (animal), tram, shuttle) + направленное движение (разные travel, repair, wend) получаются одним кластером

К сожалению, ни одним алгоритмом не выделяются четкие кластер передвижения с помощью определенного средства (типа sky, tram, swing, drive by car, ride an animal) и кластер движения по вертикальной оси (rise, accend vs fall, descend)

|сделано| баллы        | задачи          |
|-------| ------------- |-------------| 
|да| 1.5| извлечены все гипонимы синсета Х (выберите такой Х, у которого не меньше 30 гипонимов)| 
|да| 2| построен граф, где узлы -- гипонимы синсета X, а рёбра отражают колексификацию узлов|   
|да| 1| толщина ребра отражает частоту колексификации|   
|да| 0.5       | сколько получилось связных компонент?         |
|да| 1        | определён коэффициент ассортативности и плотность графа          |
|да| 2        | как распределились (взвешенные) степени узлов? какие узлы оказались центральными (попробуйте несколько метрик, например, degree centrality и eigencentrality)? прокомментируйте результаты           |
|да| 2       | разбейте граф на сообщества (поиграйте с несколькими алгоритмами) и прокомментируйте результаты         |
|да| + 1 бонусный балл       | в качестве X, взят синсет ‘travel.v.01’ (обратите внимание, что он большой, 123 гипонима)        |

*Анна Полянская, 2020*